File: internal/check/clarity.go
package check

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Nessaek/cli-a11y/internal/probe"
)

// Checks for comprehension: help you can find your way around, errors that say
// what to do next, exit codes that mean what they claim.
//
// This is the dimension people argue is "not really accessibility". It is the
// one that decides whether someone with a cognitive disability, or reading in a
// second language, or hearing the output one line at a time through speech, can
// use the tool at all.

func clarityChecks(s *probe.Set) []Finding {
	var out []Finding
	out = append(out, checkHelpExists(s))
	out = append(out, checkHelpExit(s)...)
	out = append(out, checkHelpStream(s)...)
	out = append(out, checkVersion(s)...)
	out = append(out, checkErrorExit(s)...)
	out = append(out, checkErrorStream(s)...)
	out = append(out, checkErrorActionable(s)...)
	out = append(out, checkHelpStructure(s)...)
	out = append(out, checkReadability(s)...)
	return out
}

var (
	usageRe     = regexp.MustCompile(`(?im)^\s*(usage|synopsis)\b`)
	optionsRe   = regexp.MustCompile(`(?im)^\s*(options|flags|arguments|commands|subcommands)\b`)
	dashLineRe  = regexp.MustCompile(`(?im)^\s*-{1,2}[a-z]`)
	examplesRe  = regexp.MustCompile(`(?im)^\s*(examples?)\b`)
	exampleCmd  = regexp.MustCompile(`(?m)^\s{2,}\$\s+\S`)
	versionLike = regexp.MustCompile(`(?i)\d+\.\d+|version`)
	// "Where to look next" takes many shapes: --help, a usage line, a spelling
	// suggestion, a documentation URL, or simply naming a help command. Matching
	// only the first few spellings punishes tools that do the right thing in
	// their own words.
	pointsOnRe = regexp.MustCompile(`(?i)--help\b|(?:^|\s)-h(?:\s|$)|usage:|did you mean|maybe you meant|\bhelp\b|https?://|for more (?:info|information|details)`)
)

func checkHelpExists(s *probe.Set) Finding {
	p := s.Get("help")
	if !p.OK() {
		p = s.Get("helpPipe")
	}
	if !p.OK() {
		return skip("C-HELP-EXISTS", Clarity, "Help is available", "The help invocation could not be run.")
	}
	m := s.Meta
	text := strings.TrimSpace(visibleText(p.Raw))

	// Paged help has not failed to return; it is waiting in less. Judge its
	// content by the piped run, which no pager touches.
	if p.TimedOut && isPaging(p) {
		if piped := s.Get("helpPipe"); piped.OK() {
			t := strings.TrimSpace(visibleText(piped.Raw))
			if len(t) > 20 {
				return pass("C-HELP-EXISTS", Clarity, "Help is available",
					fmt.Sprintf("%s printed %d lines, through a pager on a terminal. Paging itself is reported as V-PAGER.",
						m.HelpFlag, len(strings.Split(t, "\n"))))
			}
		}
	}

	if p.TimedOut {
		return fail("C-HELP-EXISTS", Clarity, m.HelpFlag+" does not return", Critical,
			fmt.Sprintf("Running %s did not terminate. Help is the one thing a user reaches for when stuck, and it is the first thing a screen reader user runs on an unfamiliar tool.", m.HelpFlag),
			nil, "Make the help flag short-circuit everything else and exit immediately.")
	}
	if len(text) < 20 {
		ev := []string{fmt.Sprintf("exit code %d", p.Code)}
		if text == "" {
			ev = append(ev, "no output at all")
		} else {
			ev = append(ev, "output: "+excerpt(text, 72))
		}
		return fail("C-HELP-EXISTS", Clarity, m.HelpFlag+" produces no help", Critical,
			fmt.Sprintf("Running %s produced %d characters of output. There is no way to discover what the tool does without leaving the terminal.", m.HelpFlag, len(text)),
			ev, "Print a usage summary, the available options and at least one example.")
	}
	return pass("C-HELP-EXISTS", Clarity, "Help is available",
		fmt.Sprintf("%s printed %d lines.", m.HelpFlag, len(strings.Split(text, "\n"))))
}

func checkHelpExit(s *probe.Set) []Finding {
	p := s.Get("helpPipe")
	if !p.OK() || p.TimedOut {
		return nil
	}
	if p.Code == 0 {
		return []Finding{pass("C-HELP-EXIT", Clarity, "Help exits successfully", s.Meta.HelpFlag+" exited 0.")}
	}
	return []Finding{fail("C-HELP-EXIT", Clarity, "Help exits with an error code", Minor,
		fmt.Sprintf("%s exited %d. Asking for help is not a failure, and a non-zero code here breaks shell chains and makes wrappers report a problem that did not happen.", s.Meta.HelpFlag, p.Code),
		nil, "Exit 0 when help was explicitly requested. Reserve non-zero for help printed because the invocation was wrong.")}
}

func checkHelpStream(s *probe.Set) []Finding {
	p := s.Get("helpPipe")
	if !p.OK() || p.TimedOut {
		return nil
	}
	out := strings.TrimSpace(visibleText(p.Stdout))
	errText := strings.TrimSpace(visibleText(p.Stderr))
	if len(out) > 20 {
		return []Finding{pass("C-HELP-STREAM", Clarity, "Help goes to stdout",
			"Help can be piped, paged and saved.")}
	}
	if len(errText) > 20 {
		return []Finding{fail("C-HELP-STREAM", Clarity, "Help printed to stderr", Moderate,
			fmt.Sprintf("%s writes to stderr, so \"cmd --help | less\" and \"cmd --help > notes.txt\" both come back empty. Paging is how someone reading by speech gets through a long help text at their own pace.", s.Meta.HelpFlag),
			[]string{fmt.Sprintf("stdout: %d chars, stderr: %d chars", len(out), len(errText))},
			"Requested help goes to stdout. Only help printed in response to a usage error belongs on stderr.")}
	}
	return nil
}

func checkVersion(s *probe.Set) []Finding {
	p := s.Get("version")
	if !p.OK() {
		return nil
	}
	text := strings.TrimSpace(visibleText(p.Raw))
	if p.Code == 0 && versionLike.MatchString(text) && len(text) < 400 {
		return []Finding{pass("C-VERSION", Clarity, "Reports its version", excerpt(text, 60))}
	}
	ev := []string{fmt.Sprintf("exit code %d", p.Code)}
	if text == "" {
		ev = append(ev, "no output")
	} else {
		ev = append(ev, "output: "+excerpt(text, 72))
	}
	return []Finding{fail("C-VERSION", Clarity, s.Meta.VersionFlag+" does not report a version", Minor,
		"No recognisable version string. Someone who needs help — from a colleague, a forum, or a support desk they can reach — has to start by working out which version they have.",
		ev, "Print the version and exit 0.")}
}

func checkErrorExit(s *probe.Set) []Finding {
	p := s.Get("badFlag")
	if !p.OK() || p.TimedOut {
		return nil
	}
	if p.Code != 0 {
		return []Finding{pass("C-ERROR-EXIT", Clarity, "Bad input fails loudly",
			fmt.Sprintf("An unrecognised flag exited %d.", p.Code))}
	}
	return []Finding{fail("C-ERROR-EXIT", Clarity, "Bad input exits successfully", Serious,
		fmt.Sprintf("Passing %s exited 0. A typo is silently accepted, so the only signal that something went wrong is whatever the user can spot in the output — which is no signal at all if they cannot see it, or are reading it one line at a time.", s.Meta.BadFlag),
		nil, "Reject unknown options with a non-zero exit code.")}
}

func checkErrorStream(s *probe.Set) []Finding {
	p := s.Get("badFlag")
	if !p.OK() || p.TimedOut {
		return nil
	}
	errText := strings.TrimSpace(visibleText(p.Stderr))
	out := strings.TrimSpace(visibleText(p.Stdout))
	if errText != "" {
		return []Finding{pass("C-ERROR-STREAM", Clarity, "Errors go to stderr",
			"The error is visible even when stdout is redirected.")}
	}
	if out != "" {
		return []Finding{fail("C-ERROR-STREAM", Clarity, "Errors printed to stdout", Moderate,
			"The error message went to stdout. Redirect the output to a file — which is exactly what someone does before reading it in an editor with a screen reader — and the error vanishes into the file instead of reaching the terminal.",
			[]string{"stdout: " + excerpt(out, 60)},
			"Diagnostics go to stderr; only the tool's actual product goes to stdout.")}
	}
	return []Finding{fail("C-ERROR-STREAM", Clarity, "Bad input produces no message", Serious,
		"An unrecognised flag produced no output on either stream. The user gets nothing to act on.",
		nil, "Always say what was rejected and why.")}
}

func checkErrorActionable(s *probe.Set) []Finding {
	p := s.Get("badFlag")
	if !p.OK() || p.TimedOut {
		return nil
	}
	text := strings.TrimSpace(visibleText(p.Raw))
	if text == "" {
		return nil
	}
	m := s.Meta
	lines := strings.Split(text, "\n")
	namesInput := strings.Contains(text, m.BadFlag) || strings.Contains(text, strings.TrimPrefix(m.BadFlag, "--"))
	pointsOn := pointsOnRe.MatchString(text)

	var missing []string
	if !namesInput {
		missing = append(missing, "it does not repeat back the input it rejected")
	}
	if !pointsOn {
		missing = append(missing, "it does not say where to look next")
	}
	if len(lines) > 40 {
		missing = append(missing, fmt.Sprintf("it dumps %d lines of help on top of the message", len(lines)))
	}

	first := ""
	for _, l := range lines {
		if strings.TrimSpace(l) != "" {
			first = l
			break
		}
	}
	if len(missing) == 0 {
		return []Finding{pass("C-ERROR-ACTIONABLE", Clarity, "Errors are actionable",
			"The message names the rejected input and points the user somewhere.", excerpt(first, 70))}
	}
	sev := Minor
	if len(missing) > 1 {
		sev = Moderate
	}
	return []Finding{fail("C-ERROR-ACTIONABLE", Clarity, "Error message is not actionable", sev,
		fmt.Sprintf("The message for an unrecognised flag is incomplete: %s. Read aloud, an error that does not name the offending input and does not say what to do next leaves nothing to act on.", strings.Join(missing, "; ")),
		[]string{"first line: " + excerpt(first, 70)},
		"Quote the input you rejected, suggest the nearest valid option, and name the one command that lists them all.")}
}

func checkHelpStructure(s *probe.Set) []Finding {
	help := helpText(s)
	if help == "" {
		return nil
	}
	hasUsage := usageRe.MatchString(help)
	hasOptions := optionsRe.MatchString(help) || dashLineRe.MatchString(help)
	hasExample := examplesRe.MatchString(help) || exampleCmd.MatchString(help)

	var out []Finding
	var missing []string
	if !hasUsage {
		missing = append(missing, "no usage line")
	}
	if !hasOptions {
		missing = append(missing, "no options section")
	}
	if len(missing) > 0 {
		out = append(out, fail("C-HELP-STRUCTURE", Clarity, "Help has no predictable structure", Moderate,
			fmt.Sprintf("Help text is missing landmarks users navigate by: %s. Reading by speech is linear, so a heading that can be searched for is the only way to skip to the relevant part.", strings.Join(missing, ", ")),
			nil, "Follow the conventional shape: a usage line, then a description, then options, then examples, each under a heading."))
	} else {
		out = append(out, pass("C-HELP-STRUCTURE", Clarity, "Help follows the conventional shape",
			"Usage line and options section both present."))
	}

	if hasExample {
		out = append(out, pass("C-HELP-EXAMPLES", Clarity, "Help includes examples",
			"At least one worked invocation is shown."))
	} else {
		out = append(out, fail("C-HELP-EXAMPLES", Clarity, "Help includes no examples", Minor,
			"No example invocation. An example is a template to copy; assembling one from a list of options is a working-memory task the rest of the help does not require.",
			nil, "Show two or three complete commands covering the common cases."))
	}
	return out
}

var sentenceSplit = regexp.MustCompile(`[.!?](?:\s|$)`)

func checkReadability(s *probe.Set) []Finding {
	help := helpText(s)
	if help == "" {
		return nil
	}
	lines := strings.Split(help, "\n")
	var long []string
	for _, l := range lines {
		if len(strings.TrimSpace(l)) > 100 {
			long = append(long, l)
		}
	}
	words := strings.Fields(help)
	var sentences []string
	for _, sent := range sentenceSplit.Split(help, -1) {
		if len(strings.Fields(sent)) > 3 {
			sentences = append(sentences, sent)
		}
	}
	avg := 0.0
	if len(sentences) > 0 {
		total := 0
		for _, sent := range sentences {
			total += len(strings.Fields(sent))
		}
		avg = float64(total) / float64(len(sentences))
	}

	var problems []string
	if len(long) > 3 {
		problems = append(problems, fmt.Sprintf("%d lines run past 100 characters", len(long)))
	}
	if avg > 25 {
		problems = append(problems, fmt.Sprintf("sentences average %.0f words", avg))
	}
	if len(lines) > 250 {
		problems = append(problems, fmt.Sprintf("%d lines of help arrive in one block", len(lines)))
	}

	if len(problems) == 0 {
		return []Finding{pass("C-READABILITY", Clarity, "Help is readable",
			fmt.Sprintf("%d lines, %d words, sentences averaging %.0f words.", len(lines), len(words), avg))}
	}
	var ev []string
	for _, l := range long {
		ev = append(ev, fmt.Sprintf("%d chars: %s", len(strings.TrimSpace(l)), excerpt(l, 60)))
	}
	return []Finding{fail("C-READABILITY", Clarity, "Help is hard to read", Minor,
		fmt.Sprintf("%s. Long lines and long sentences cost the most for readers with dyslexia or a cognitive disability, and for anyone hearing the text rather than scanning it.", strings.Join(problems, "; ")),
		trim(ev, 3),
		"Wrap to the terminal width, keep sentences short, and split a long help into subcommand help pages.")}
}

File: src/checks/clarity.js
'use strict';

// Checks for comprehension: help you can find your way around, errors that say
// what to do next, exit codes that mean what they claim.
//
// This is the dimension people argue is "not really accessibility". It is the
// one that decides whether someone with a cognitive disability, or reading in a
// second language, or hearing the output one line at a time through speech,
// can use the tool at all.

const U = require('./util');

const DIM = 'clarity';

function checkHelpExists(probes) {
  const p = probes.help || probes.helpPipe;
  const meta = probes._meta;
  if (!U.ok(p)) return [];
  const text = U.visibleText(p.raw).trim();

  // Paged help has not failed to return; it is waiting in less. Judge its
  // content by the piped run, which no pager touches.
  if (p.timedOut && U.isPaging(p)) {
    const piped = U.visibleText((probes.helpPipe && probes.helpPipe.raw) || '').trim();
    if (piped.length > 20) {
      return [U.pass({
        id: 'C-HELP-EXISTS', dimension: DIM, title: 'Help is available',
        detail: `${meta.helpFlag} printed ${piped.split('\n').length} lines, through a pager on a terminal. Paging itself is reported as V-PAGER.`,
      })];
    }
  }

  if (p.timedOut) {
    return [U.fail({
      id: 'C-HELP-EXISTS',
      dimension: DIM,
      title: `${meta.helpFlag} does not return`,
      severity: 'critical',
      detail: `Running ${meta.helpFlag} did not terminate. Help is the one thing a user reaches for when stuck, and it is the first thing a screen reader user runs on an unfamiliar tool.`,
      remedy: 'Make the help flag short-circuit everything else and exit immediately.',
    })];
  }
  if (text.length < 20) {
    return [U.fail({
      id: 'C-HELP-EXISTS',
      dimension: DIM,
      title: `${meta.helpFlag} produces no help`,
      severity: 'critical',
      detail: `Running ${meta.helpFlag} produced ${text.length} characters of output. There is no way to discover what the tool does without leaving the terminal.`,
      evidence: [`exit code ${p.code}`, text ? `output: ${U.excerpt(text)}` : 'no output at all'],
      remedy: 'Print a usage summary, the available options and at least one example.',
    })];
  }
  return [U.pass({
    id: 'C-HELP-EXISTS', dimension: DIM, title: 'Help is available',
    detail: `${meta.helpFlag} printed ${text.split('\n').length} lines.`,
  })];
}

function checkHelpExit(probes) {
  const p = probes.helpPipe;
  const meta = probes._meta;
  if (!U.ok(p) || p.timedOut) return [];
  if (p.code === 0) {
    return [U.pass({
      id: 'C-HELP-EXIT', dimension: DIM, title: 'Help exits successfully',
      detail: `${meta.helpFlag} exited 0.`,
    })];
  }
  return [U.fail({
    id: 'C-HELP-EXIT',
    dimension: DIM,
    title: 'Help exits with an error code',
    severity: 'minor',
    detail: `${meta.helpFlag} exited ${p.code}. Asking for help is not a failure, and a non-zero code here breaks shell chains and makes wrappers report a problem that did not happen.`,
    remedy: 'Exit 0 when help was explicitly requested. Reserve non-zero for help printed because the invocation was wrong.',
  })];
}

function checkHelpStream(probes) {
  const p = probes.helpPipe;
  const meta = probes._meta;
  if (!U.ok(p) || p.timedOut) return [];
  const out = U.visibleText(p.stdout).trim();
  const err = U.visibleText(p.stderr).trim();
  if (out.length > 20) {
    return [U.pass({
      id: 'C-HELP-STREAM', dimension: DIM, title: 'Help goes to stdout',
      detail: 'Help can be piped, paged and saved.',
    })];
  }
  if (err.length > 20) {
    return [U.fail({
      id: 'C-HELP-STREAM',
      dimension: DIM,
      title: 'Help printed to stderr',
      severity: 'moderate',
      detail: `${meta.helpFlag} writes to stderr, so "cmd --help | less" and "cmd --help > notes.txt" both come back empty. Paging is how someone reading by speech gets through a long help text at their own pace.`,
      evidence: [`stdout: ${out.length} chars, stderr: ${err.length} chars`],
      remedy: 'Requested help goes to stdout. Only help printed in response to a usage error belongs on stderr.',
    })];
  }
  return [];
}

function checkVersion(probes) {
  const p = probes.version;
  const meta = probes._meta;
  if (!U.ok(p)) return [];
  const text = U.visibleText(p.raw).trim();
  const looksVersioned = /\d+\.\d+/.test(text) || /version/i.test(text);
  if (p.code === 0 && looksVersioned && text.length < 400) {
    return [U.pass({
      id: 'C-VERSION', dimension: DIM, title: 'Reports its version',
      detail: U.excerpt(text, 60),
    })];
  }
  return [U.fail({
    id: 'C-VERSION',
    dimension: DIM,
    title: `${meta.versionFlag} does not report a version`,
    severity: 'minor',
    detail: 'No recognisable version string. Someone who needs help — from a colleague, a forum, or a support desk they can reach — has to start by working out which version they have.',
    evidence: [`exit code ${p.code}`, text ? `output: ${U.excerpt(text)}` : 'no output'],
    remedy: 'Print the version and exit 0.',
  })];
}

function checkErrorExit(probes) {
  const p = probes.badFlag;
  const meta = probes._meta;
  if (!U.ok(p) || p.timedOut) return [];
  if (p.code !== 0) {
    return [U.pass({
      id: 'C-ERROR-EXIT', dimension: DIM, title: 'Bad input fails loudly',
      detail: `An unrecognised flag exited ${p.code}.`,
    })];
  }
  return [U.fail({
    id: 'C-ERROR-EXIT',
    dimension: DIM,
    title: 'Bad input exits successfully',
    severity: 'serious',
    detail: `Passing ${meta.badFlag} exited 0. A typo is silently accepted, so the only signal that something went wrong is whatever the user can spot in the output — which is no signal at all if they cannot see it, or are reading it one line at a time.`,
    remedy: 'Reject unknown options with a non-zero exit code.',
  })];
}

function checkErrorStream(probes) {
  const p = probes.badFlag;
  if (!U.ok(p) || p.timedOut) return [];
  const err = U.visibleText(p.stderr).trim();
  const out = U.visibleText(p.stdout).trim();
  if (err.length > 0) {
    return [U.pass({
      id: 'C-ERROR-STREAM', dimension: DIM, title: 'Errors go to stderr',
      detail: 'The error is visible even when stdout is redirected.',
    })];
  }
  if (out.length > 0) {
    return [U.fail({
      id: 'C-ERROR-STREAM',
      dimension: DIM,
      title: 'Errors printed to stdout',
      severity: 'moderate',
      detail: 'The error message went to stdout. Redirect the output to a file — which is exactly what someone does before reading it in an editor with a screen reader — and the error vanishes into the file instead of reaching the terminal.',
      evidence: [`stdout: ${U.excerpt(out, 60)}`],
      remedy: 'Diagnostics go to stderr; only the tool\'s actual product goes to stdout.',
    })];
  }
  return [U.fail({
    id: 'C-ERROR-STREAM',
    dimension: DIM,
    title: 'Bad input produces no message',
    severity: 'serious',
    detail: 'An unrecognised flag produced no output on either stream. The user gets nothing to act on.',
    remedy: 'Always say what was rejected and why.',
  })];
}

function checkErrorActionable(probes) {
  const p = probes.badFlag;
  const meta = probes._meta;
  if (!U.ok(p) || p.timedOut) return [];
  const text = U.visibleText(p.raw).trim();
  if (!text) return [];

  const namesInput = text.includes(meta.badFlag) || text.includes(meta.badFlag.replace(/^--/, ''));
  // "Where to look next" takes many shapes: --help, a usage line, a spelling
  // suggestion, a documentation URL, or simply naming a help command. Matching
  // only the first few spellings punishes tools that do the right thing in
  // their own words.
  const pointsOn = /--help\b|(?:^|\s)-h(?:\s|$)|usage:|did you mean|maybe you meant|\bhelp\b|https?:\/\/|for more (?:info|information|details)/i
    .test(text);
  const wall = text.split('\n').length > 40;

  const missing = [];
  if (!namesInput) missing.push('it does not repeat back the input it rejected');
  if (!pointsOn) missing.push('it does not say where to look next');
  if (wall) missing.push(`it dumps ${text.split('\n').length} lines of help on top of the message`);

  if (!missing.length) {
    return [U.pass({
      id: 'C-ERROR-ACTIONABLE', dimension: DIM, title: 'Errors are actionable',
      detail: 'The message names the rejected input and points the user somewhere.',
      evidence: [U.excerpt(text.split('\n').find((l) => l.trim()) || '', 70)],
    })];
  }

  return [U.fail({
    id: 'C-ERROR-ACTIONABLE',
    dimension: DIM,
    title: 'Error message is not actionable',
    severity: missing.length > 1 ? 'moderate' : 'minor',
    detail: `The message for an unrecognised flag is incomplete: ${missing.join('; ')}. Read aloud, an error that does not name the offending input and does not say what to do next leaves nothing to act on.`,
    evidence: [`first line: ${U.excerpt(text.split('\n').find((l) => l.trim()) || '', 70)}`],
    remedy: 'Quote the input you rejected, suggest the nearest valid option, and name the one command that lists them all.',
  })];
}

function checkHelpStructure(probes) {
  const help = U.helpText(probes);
  if (!help) return [];
  const hasUsage = /^\s*(usage|synopsis)\b/im.test(help);
  const hasOptions = /^\s*(options|flags|arguments|commands|subcommands)\b/im.test(help)
    || /^\s*-{1,2}[a-z]/im.test(help);
  const hasExample = /^\s*(examples?)\b/im.test(help)
    || /^\s{2,}\$\s+\S/m.test(help);

  const missing = [];
  if (!hasUsage) missing.push('no usage line');
  if (!hasOptions) missing.push('no options section');

  const findings = [];
  if (missing.length) {
    findings.push(U.fail({
      id: 'C-HELP-STRUCTURE',
      dimension: DIM,
      title: 'Help has no predictable structure',
      severity: 'moderate',
      detail: `Help text is missing landmarks users navigate by: ${missing.join(', ')}. Reading by speech is linear, so a heading that can be searched for is the only way to skip to the relevant part.`,
      remedy: 'Follow the conventional shape: a usage line, then a description, then options, then examples, each under a heading.',
    }));
  } else {
    findings.push(U.pass({
      id: 'C-HELP-STRUCTURE', dimension: DIM, title: 'Help follows the conventional shape',
      detail: 'Usage line and options section both present.',
    }));
  }

  findings.push(hasExample
    ? U.pass({
      id: 'C-HELP-EXAMPLES', dimension: DIM, title: 'Help includes examples',
      detail: 'At least one worked invocation is shown.',
    })
    : U.fail({
      id: 'C-HELP-EXAMPLES',
      dimension: DIM,
      title: 'Help includes no examples',
      severity: 'minor',
      detail: 'No example invocation. An example is a template to copy; assembling one from a list of options is a working-memory task the rest of the help does not require.',
      remedy: 'Show two or three complete commands covering the common cases.',
    }));

  return findings;
}

function checkReadability(probes) {
  const help = U.helpText(probes);
  if (!help) return [];
  const lines = help.split('\n');
  const longLines = lines.filter((l) => l.trim().length > 100);
  const words = help.split(/\s+/).filter(Boolean);
  const sentences = help.split(/[.!?](?:\s|$)/).filter((s) => s.trim().split(/\s+/).length > 3);
  const avgSentence = sentences.length
    ? sentences.reduce((a, s) => a + s.trim().split(/\s+/).length, 0) / sentences.length
    : 0;

  const problems = [];
  if (longLines.length > 3) {
    problems.push(`${longLines.length} lines run past 100 characters`);
  }
  if (avgSentence > 25) {
    problems.push(`sentences average ${avgSentence.toFixed(0)} words`);
  }
  if (lines.length > 250) {
    problems.push(`${lines.length} lines of help arrive in one block`);
  }

  if (!problems.length) {
    return [U.pass({
      id: 'C-READABILITY', dimension: DIM, title: 'Help is readable',
      detail: `${lines.length} lines, ${words.length} words, sentences averaging ${avgSentence.toFixed(0)} words.`,
    })];
  }

  return [U.fail({
    id: 'C-READABILITY',
    dimension: DIM,
    title: 'Help is hard to read',
    severity: 'minor',
    detail: `${problems.join('; ')}. Long lines and long sentences cost the most for readers with dyslexia or a cognitive disability, and for anyone hearing the text rather than scanning it.`,
    evidence: longLines.slice(0, 3).map((l) => `${l.trim().length} chars: ${U.excerpt(l, 60)}`),
    remedy: 'Wrap to the terminal width, keep sentences short, and split a long help into subcommand help pages.',
  })];
}

function checks(probes) {
  return [
    ...checkHelpExists(probes),
    ...checkHelpExit(probes),
    ...checkHelpStream(probes),
    ...checkVersion(probes),
    ...checkErrorExit(probes),
    ...checkErrorStream(probes),
    ...checkErrorActionable(probes),
    ...checkHelpStructure(probes),
    ...checkReadability(probes),
  ];
}

module.exports = { checks };
